File: backend/src/main/java/com/devmatch/backend/domain/project/entity/Project.java
package com.devmatch.backend.domain.project.entity;

import static jakarta.persistence.FetchType.LAZY;

import com.devmatch.backend.domain.application.entity.Application;
import com.devmatch.backend.domain.application.enums.ApplicationStatus;
import com.devmatch.backend.domain.user.entity.User;
import jakarta.persistence.*;
import java.time.LocalDateTime;
import java.util.List;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Getter
@NoArgsConstructor
@Entity
@Table(
    name = "projects",
    indexes = {@Index(name = "idx_creator_id", columnList = "creator_id")}
)
public class Project {

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  private Long id;

  private String title;
  private String description;
  private String techStack;

  private Integer teamSize;
  private Integer currentTeamSize;

  @ManyToOne
  @JoinColumn(name = "creator_id")
  private User creator;

  @Enumerated(EnumType.STRING)
  private ProjectStatus status;

  private String content;
  private Integer durationWeeks;
  private LocalDateTime createdAt;

  @OneToMany(mappedBy = "project", fetch = LAZY, orphanRemoval = true)
  private List<Application> applications;

  public Project(
      String title,
      String description,
      String techStack,
      Integer teamSize,
      User creator,
      Integer durationWeeks
  ) {
    this.title = title;
    this.description = description;
    this.techStack = techStack;
    this.teamSize = teamSize;
    this.creator = creator;
    this.status = ProjectStatus.RECRUITING;
    this.currentTeamSize = 0;
    this.content = "";
    this.durationWeeks = durationWeeks;
    this.createdAt = LocalDateTime.now();
  }

  public void changeStatus(ProjectStatus newStatus) {
    if (newStatus == this.status) {
      throw new IllegalArgumentException(
          "현재 상태(%s)와 동일한 상태(%s)로 변경할 수 없습니다".formatted(this.status, newStatus));
    }

    this.status = newStatus;
  }

  public void changeCurTeamSize(ApplicationStatus oldStatus, ApplicationStatus newStatus) {
    if (oldStatus != ApplicationStatus.APPROVED && newStatus == ApplicationStatus.APPROVED) {
      if (this.currentTeamSize.equals(this.teamSize)) {
        throw new IllegalArgumentException("정원이 가득 차서 지원서를 더 이상 승인할 수 없습니다");
      }

      this.currentTeamSize++;
    } else if (oldStatus == ApplicationStatus.APPROVED && newStatus != ApplicationStatus.APPROVED) {
      this.currentTeamSize--;
    }

    if (this.currentTeamSize.equals(this.teamSize)) {
      this.status = ProjectStatus.COMPLETED;
    } else {
      this.status = ProjectStatus.RECRUITING;
    }
  }

  public void changeContent(String content) {
    this.content = content;
  }
}

File: backend/src/main/java/com/devmatch/backend/domain/project/entity/Project.kt
package com.devmatch.backend.domain.project.entity

import com.devmatch.backend.domain.application.entity.Application
import com.devmatch.backend.domain.user.entity.User
import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "projects", indexes = [Index(name = "idx_creator_id", columnList = "creator_id")])
class Project(
    val title: String,
    val description: String,
    val techStack: String,
    val teamSize: Int,
    @ManyToOne(fetch = FetchType.LAZY) @JoinColumn(name = "creator_id") val creator: User,
    val durationWeeks: Int
) {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long? = null

    @Enumerated(EnumType.STRING)
    var status: ProjectStatus = ProjectStatus.RECRUITING

    var content = ""
    var currentTeamSize = 0
    val createdAt: LocalDateTime = LocalDateTime.now()

    @OneToMany(mappedBy = "project", fetch = FetchType.LAZY, orphanRemoval = true)
    private val applications: List<Application> = emptyList()

    fun changeStatus(newStatus: ProjectStatus) {
        require(newStatus != status) { "현재 상태(${status})와 동일한 상태(${newStatus})로 변경할 수 없습니다" }
        status = newStatus
    }

    fun increaseCurTeamSize() {
        require(currentTeamSize < teamSize) { "정원이 가득 차서 지원서를 더 이상 승인할 수 없습니다" }
        currentTeamSize++

        if (currentTeamSize == teamSize) {
            status = ProjectStatus.COMPLETED
        }
    }

    fun decreaseCurTeamSize() {
        currentTeamSize--
        status = ProjectStatus.RECRUITING
    }
}